Support interface.macAddress

booxter opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: support interface.macAddress field for pod network types supported so far (slirp and bridge).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #1069
Fixes #1251

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Support interface.macAddress field to override MAC address exposed to the guest.
```

Documentation update: https://github.com/kubevirt/user-guide/pull/94

booxter commented on pkg/virt-api/validating-webhook/validating-webhook_test.go:
```diff
@@ -811,6 +811,28 @@ var _ = Describe("Validating Webhook", func() {
+			for _, macAddress := range []string{"de:ad:00:00:be", "de-ad-00-00-be-af"} {
+				vmi.Spec.Domain.Devices.Interfaces[0].MacAddress = macAddress // missing octet
+				causes := ValidateVirtualMachineInstanceSpec(k8sfield.NewPath("fake"), &vmi.Spec)
+				// if this is processed correctly, it should not result in any error
```
to remove

booxter commented on tests/utils.go:
```diff
@@ -991,6 +991,14 @@ func NewRandomVMIWithe1000NetworkInterface() *v1.VirtualMachineInstance {
+	// Use alpine because cirros dhcp client starts prematurily before link is ready
+	vmi := NewRandomVMIWithEphemeralDisk(RegistryDiskFor(RegistryDiskAlpine))
+	AddExplicitPodNetworkInterface(vmi)
+	vmi.Spec.Domain.Devices.Interfaces[0].MacAddress = "de:ad:00:00:be:af"
```
put the literal into a constant and reuse in the func test

booxter commented on tests/utils.go:
```diff
@@ -991,6 +991,14 @@ func NewRandomVMIWithe1000NetworkInterface() *v1.VirtualMachineInstance {
 }
 
+func NewRandomVMIWithCustomMacAddress() *v1.VirtualMachineInstance {
+	// Use alpine because cirros dhcp client starts prematurily before link is ready
```
not applicable here, it's comment for e1000 only

booxter commented:
/hold

booxter commented:
/hold cancel

booxter commented on pkg/virt-api/validating-webhook/validating-webhook.go:
```diff
@@ -47,6 +47,10 @@ const (
 
 var validInterfaceModels = []string{"e1000", "e1000e", "ne2k_pci", "pcnet", "rtl8139", "virtio"}
+var macAddressRegex = regexp.MustCompile(
+	// there are other MAC schemes (e.g. with dashes instead of colons),
```
it's actually arguable if we want just colons-form. Chime in with comments if you think we should support different forms.

booxter replied:
@fabiand what do you think about supporting other MAC address formats known to golang net package? (Like DE-AD-00-00-BE-AF.) Do we want to support them? If so, do we store them as-is or convert to a canonical form? If not, do we want to make the rules even more strict, f.e. forcing lower-case letters?

SchSeba replied:
I think we need to support any mac address format just call the net.ParseMAC and check for error.

booxter commented:
@vladikr @SchSeba please take a look, this PR should be ready to consume in master.

booxter commented:
please retest this

rmohr commented:
retest this please

vladikr commented:
Looks good to me. I suppose we could address the issue with the replica set later, when we will support mac pools?

booxter commented:
Yes, discussion about replica sets probably belongs to a follow-up PR once we have an idea on how we allocate unique MAC addresses (and whether we enforce their uniqueness / persistence). There could be a (wild) use case to share MAC addresses.

booxter commented:
The latest upload supports alternative MAC formats in API (and validates that at least some of them actually work and correctly propagated into the guest OS; I haven't added functional tests for ALL of supported formats though because it would blow up time of execution of the suite even further).

booxter commented:
The latest upload forbids non -48 MAC address formats though: I checked and -64 doesn't work at all (qemu fails to start.)

davidvossel commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -111,8 +112,20 @@ func getBinding(iface *v1.Interface, domain *api.Domain) (BindMechanism, error)
 		return nil, err
 	}
 
+	populateMacAddress := func(vif *VIF, iface *v1.Interface) error {
```
is this logic covered in the podinterface_test.go unit test?

davidvossel commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -311,6 +326,10 @@ func (s *SlirpPodInterface) preparePodNetworkInterfaces() error {
 
 func (s *SlirpPodInterface) decorateConfig() error {
 	s.domain.Spec.QEMUCmd.QEMUArg[s.podInterfaceNum].Value += fmt.Sprintf(",id=%s", s.iface.Name)
+	if s.iface.MacAddress != "" {
```
I'm not seeing unit test cases for this.

davidvossel commented on tests/vmi_networking_test.go:
```diff
@@ -108,11 +108,14 @@ var _ = Describe("Networking", func() {
 					Expect(networkVMI.Spec.Domain.Devices.Interfaces).ToNot(BeZero())
 					Expect(networkVMI.Spec.Networks).ToNot(BeZero())
+					// also set custom MAC address to the interface
+					networkVMI.Spec.Domain.Devices.Interfaces[0].MacAddress = fmt.Sprintf("de:ad:00:0%d:be:af", i)
```
with this addition, do we have any functional tests that verify connectivity that don't explicitly set the mac now?

booxter replied:
No, though we have checks for "implicit" case (no interfaces in spec, so default one is used); I will reinstate it, but consider that the test matrix starts to blow up if we add all connectivity tests to each variant of configuration. Test run time is not ideal already, I was thinking maybe doing these test matrix multiplications is problematic.

davidvossel commented on tests/vmi_networking_test.go:
```diff
@@ -402,4 +405,51 @@ var _ = Describe("Networking", func() {
+		})
+	})
+
+	Context("VirtualMachineInstance with custom MAC address and slirp interface", func() {
```
Isn't this test case already covered by the slirp functional tests?

booxter replied:
No, we don't check that correct MAC address is set inside the guest there.

yuvalif commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -111,8 +112,20 @@ func getBinding(iface *v1.Interface, domain *api.Domain) (BindMechanism, error)
+
 	if iface.Bridge != nil {
 		vif := &VIF{Name: podInterface}
+		populateMacAddress(vif, iface)
```
can't we set the mac address in case of slirp?

booxter replied:
We can but we don't need to proxy it through our struct and just extract it from v1.interface when constructing qemu args.

booxter replied:
@yuvalif see https://github.com/kubevirt/kubevirt/pull/1213/commits/d61c1671037fef382a71181d407d68aa220cae25

yuvalif replied:
ok. got it

booxter commented:
The latest upload adds missing unit tests; also restores functional connectivity tests for the case of explicit pod network definition that reuses the default MAC address.

yuvalif commented on pkg/virt-launcher/virtwrap/network/podinterface.go:
```diff
@@ -153,13 +166,15 @@ func (b *BridgePodInterface) discoverPodNetworkInterface() error {
-	if err != nil {
-		log.Log.Reason(err).Errorf("failed to get MAC for %s", podInterface)
-		return err
+	if len(b.vif.MAC) == 0 {
```
ok. i see where the concern is. if some network plugin manages the mac addresses somehow.
why did we originally move the mac from the pod interface to the VM? couldn't we just send an arp notifying the IP address now belong to a new mac?

booxter replied:
No idea really. :) @vladikr @phoracek any back story behind retaining MAC address in the initial podinterface.go implementation?

vladikr replied:
We just wanted to have a generic solution that doesn't rely on the CNI network provider. Some providers use DHCP to hand out IPs, some just automagically set it. We wanted to be sure.

I'm not 100% confident that this solution will work with all providers, nor whether it will work over a long period of time (if the vm restarts after the original lease from CNI DHCP is expired) - however, I know that we have another solution that worked so far.
I'm comfortable enough to merge this.

booxter commented:
Note: I may need to rework functional tests in light of: https://github.com/kubevirt/kubevirt/pull/1240

booxter commented:
/hold

booxter commented:
Hm shouldn't @kubevirt-bot have added a do-not-merge/hold tag?..

rmohr commented:
@booxter found an error in the logs:

```
{"author":"booxter","component":"hook","error":"failed to get the labels on kubevirt/kubevirt#1213: sleep time for token reset exceeds max sleep time (22m11.101836998s \u003e 2m0s)","event-GUID":"86343060-7afa-11e8-80cb-968fd11cc030","event-type":"issue_comment","level":"error","msg":"Error handling GenericCommentEvent.","org":"kubevirt","plugin":"hold","pr":1213,"repo":"kubevirt","time":"2018-06-28T17:41:44Z","url":"https://github.com/kubevirt/kubevirt/pull/1213#issuecomment-401116419"}
```

Should be good enough to just write `/hold` again if that happens again. I will check if it is possible to  configure retries.

fabiand commented:
retest this please

booxter commented:
Rebased due to git conflicts; adopted functional tests to latest changes; added test case to validate http fetch working for all VMIs (in addition to ICMP).

booxter commented:
/hold

booxter commented:
/hold

booxter commented:
For the record, there is no longer a reason to hold it; it's ready to review / merge (I hope).

rmohr commented:
/hold cancel

booxter commented:
The latest upload migrates slirp functional tests to tables, per @rmohr request.

vladikr commented:
ok to test

vladikr commented:
All previous comments were addressed and the PR looks good to me!
Let's just wait for the tests to pass before merging.

rmohr commented on tests/vmi_slirp_interface_test.go:
```diff
@@ -42,67 +43,74 @@ var _ = Describe("Slirp", func() {
+		log.Log.Infof("%v", out)
+		Expect(err).ToNot(HaveOccurred())
+	},
+		table.Entry("VirtualMachineInstance with slirp interface", &genericVmi),
```
I meant multiple tables with two entries intead of wrapping all tests but works for me also this way.

vladikr commented:
retest this please

vladikr commented:
Tests are passing, besides one on the openshift-3.10-release, but it seems to be related to a timeout, as the rest of the failed tests in this job.
